File: SocketServer.Crypto/ServerAuthority.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Security;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Generators;
using Org.BouncyCastle.X509;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Asn1.X509;

namespace SocketServer.Crypto
{
    public class ServerAuthority
    {
        private readonly DHParameters parameters;
        private readonly AsymmetricCipherKeyPair kp;
        private readonly IBasicAgreement agreement;

        public ServerAuthority(BigInteger prime, BigInteger g) 
            : this(new DHParameters(prime, g))
        {
        }

        public ServerAuthority(int bitLength, int probability)
            : this(GetDiffieHellmanParameters(bitLength, probability))
        {
        }

        public ServerAuthority(DHParameters parameters)
        {
            this.parameters = parameters;

            IAsymmetricCipherKeyPairGenerator keyGen = GeneratorUtilities.GetKeyPairGenerator("DH");
            KeyGenerationParameters kgp = new DHKeyGenerationParameters(new SecureRandom(), parameters);

            keyGen.Init(kgp);
            kp = keyGen.GenerateKeyPair();
            agreement = AgreementUtilities.GetBasicAgreement("DH");
            agreement.Init(kp.Private);
        }


        public AsymmetricCipherKeyPair KeyPair
        {
            get { return kp;  }
        }
    
        public DHParameters Parameters
        {
            get { return parameters;  }
        }
    

        public BigInteger G
        {
            get { return parameters.G; }
        }

        public BigInteger P
        {
            get { return parameters.P; }
        }

        private static DHParameters GetDiffieHellmanParameters(int bitLength, int probability)
        {
            DHParametersGenerator generator = new DHParametersGenerator();
            generator.Init(bitLength, probability, new SecureRandom());
            return generator.GenerateParameters();
        }

        public AsymmetricKeyParameter GetPublicKeyParameter()
        {
            return kp.Public;
        }

        public byte [] GenerateEncodedPublicKeyInfo()
        {
            SubjectPublicKeyInfo publicKeyInfo 
                = SubjectPublicKeyInfoFactory
                    .CreateSubjectPublicKeyInfo(GetPublicKeyParameter());

            return publicKeyInfo.ToAsn1Object().GetDerEncoded();
        }

        public BigInteger GenerateAgreementValue(AsymmetricKeyParameter remotePublicKey)
        {
            return agreement.CalculateAgreement(remotePublicKey);
        }

    }
}

File: Java/BlazeServer/src/com/BlazeServer/Crypto/ServerAuthority.java
package com.BlazeServer.Crypto;

import java.math.BigInteger;
import java.security.SecureRandom;

import org.bouncycastle.asn1.DEREncodable;
import org.bouncycastle.asn1.DERInteger;
import org.bouncycastle.asn1.pkcs.DHParameter;
import org.bouncycastle.asn1.pkcs.PKCSObjectIdentifiers;
import org.bouncycastle.asn1.x509.AlgorithmIdentifier;
import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
import org.bouncycastle.crypto.KeyGenerationParameters;
import org.bouncycastle.crypto.agreement.DHBasicAgreement;
import org.bouncycastle.crypto.generators.DHBasicKeyPairGenerator;
import org.bouncycastle.crypto.generators.DHParametersGenerator;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.crypto.params.DHKeyGenerationParameters;
import org.bouncycastle.crypto.params.DHParameters;
import org.bouncycastle.crypto.params.DHPublicKeyParameters;

public class ServerAuthority {

	private final DHParameters parameters;
	private final AsymmetricCipherKeyPair keyPair;
	private final DHBasicAgreement agreement;
	
	public ServerAuthority(BigInteger prime, BigInteger g)  {
		this(new DHParameters(prime, g));
	}
	
	public ServerAuthority(int bitLength, int primeProbability) {
		this(getDiffieHellmanParameters(bitLength, primeProbability));
	}

	public ServerAuthority(DHParameters parameters) {
		this.parameters = parameters;
		
        DHBasicKeyPairGenerator keyPairGenerator = new DHBasicKeyPairGenerator();
        KeyGenerationParameters keyGenerationParameters = new DHKeyGenerationParameters(new SecureRandom(), parameters);
        keyPairGenerator.init(keyGenerationParameters);
        
        this.keyPair = keyPairGenerator.generateKeyPair();
        this.agreement = new DHBasicAgreement();
        this.agreement.init(keyPair.getPrivate());
	}
	
	public DHParameters getParameters() {
		return this.parameters;
	}
	
	public AsymmetricCipherKeyPair getKeyPair() {
		return this.keyPair;
	}
	
	public BigInteger getP() {
		return this.parameters.getP();
	}
	
	public BigInteger getG() {
		return this.parameters.getG();
	}
	
	public AsymmetricKeyParameter getPublicKeyParameter() {
		return (AsymmetricKeyParameter) this.keyPair.getPublic();
	}
	
    public byte [] generateEncodedPublicKeyInfo()
    {
        DHPublicKeyParameters key = (DHPublicKeyParameters) keyPair.getPublic();
		DHParameters kp = key.getParameters();
    	
		SubjectPublicKeyInfo keyInfo = new SubjectPublicKeyInfo(
            new AlgorithmIdentifier(
				PKCSObjectIdentifiers.dhKeyAgreement,
				new DHParameter(kp.getP(), kp.getG(), kp.getL()).toASN1Object()),
				(DEREncodable) new DERInteger(key.getY()));   

        return keyInfo.toASN1Object().getDEREncoded();
    }

    public BigInteger generateAgreementValue(AsymmetricKeyParameter remotePublicKey)
    {
        return agreement.calculateAgreement(remotePublicKey);
    }	
	
    private static DHParameters getDiffieHellmanParameters(int bitLength, int probability)
    {
        DHParametersGenerator generator = new DHParametersGenerator();
        generator.init(bitLength, probability, new SecureRandom());
        return generator.generateParameters();
    }	
	
}
